handlers_test.go

```diff
@@ -91,3 +92,111 @@ func Test_GivenCorrectId_HandlerReturnsDeviceObject(t *testing.T) {
+	mockServer := httptest.NewServer(r)
+
+	urls := []string{"/devices?limit=a", "/devices?page=a",
+		"/devices?limit=1&page=a", "/devices?limit=-1",
+		"/devices?page=-1"}
+	for _, url := range urls {
+		resp, err := http.Get(mockServer.URL + url)
+
+		assert.NoError(t, err)
+		assert.Equal(t, http.StatusBadRequest, resp.StatusCode)
+	}
+}
+
+func Test_GetAllDevicesHandler_NoParams_HandlerDefaultsLimitAndPage(t *testing.T) {
+	r := newRouter(NewService(&mockDao{}))
+	mockServer := httptest.NewServer(r)
+
+	resp, err := http.Get(mockServer.URL + "/devices")
+
+	assert.NoError(t, err)
+	assert.Equal(t, http.StatusOK, resp.StatusCode)
+}
+
+func Test_GetAllDevicesHandler_GivenDaoError_HandlerReturns500(t *testing.T) {
+	r := newRouter(NewService(&mockDao{returnErr: ErrDao("")}))
+	mockServer := httptest.NewServer(r)
+
+	resp, err := http.Get(mockServer.URL + "/devices")
+
+	assert.NoError(t, err)
+	assert.Equal(t, http.StatusInternalServerError, resp.StatusCode)
+}
+
+func Test_GetAllDevicesHandler_GivenLimitZero_HandlerReturnsAllDevices(t *testing.T) {
+	m := &mockDao{data: make(map[int]Device)}
+	m.data[0] = Device{Name: "test name"}
+	r := newRouter(NewService(m))
+	mockServer := httptest.NewServer(r)
+
+	resp, err := http.Get(mockServer.URL + "/devices?limit=0")
+	assert.NoError(t, err)
+	defer resp.Body.Close()
+
+	expected := []Device{{Name: "test name"}}
+
+	var result []Device
+	err = json.NewDecoder(resp.Body).Decode(&result)
+
+	fmt.Println(result)
+
+	assert.NoError(t, err)
+	assert.Equal(t, expected, result)
+}
+
+func Test_GetAllDevicesHandler_GivenPageThatHasNoDevicesToShow_HandlerReturnsEmptyJsonArray(t *testing.T) {
+	r := newRouter(NewService(&mockDao{}))
+	mockServer := httptest.NewServer(r)
+
+	resp, err := http.Get(mockServer.URL + "/devices?page=0")
+	assert.NoError(t, err)
+	defer resp.Body.Close()
+
+	expected := []int{}
+
+	var result []int
+	err = json.NewDecoder(resp.Body).Decode(&result)
+
+	assert.NoError(t, err)
+	assert.Equal(t, expected, result)
+}
+
+func Test_GetAllDevicesHandler_GivenLimitAndPage_HandlerReturnsCorrectList(t *testing.T) {
+	m := &mockDao{data: make(map[int]Device)}
+	for _, i := range []int{1, 3, 5, 2, 4} {
+		m.data[i] = Device{Id: i}
+	}
+	r := newRouter(NewService(m))
+	mockServer := httptest.NewServer(r)
+
+	urls := []string{"/devices?limit=2&page=1", "/devices?limit=2&page=2",
+		"/devices", "/devices?limit=1&page=4",
+		"/devices?limit=3"}
+
+	expected := [][]Device{
+		{Device{Id: 3}, Device{Id: 4}},
+		{Device{Id: 5}},
+		{Device{Id: 1}, Device{Id: 2}, Device{Id: 3}, Device{Id: 4}, Device{Id: 5}},
+		{Device{Id: 5}},
+		{Device{Id: 1}, Device{Id: 2}, Device{Id: 3}}}
+
+	var result []Device
+
+	for i, url := range urls {
+		resp, err := http.Get(mockServer.URL + url)
+		assert.NoError(t, err)
+		defer resp.Body.Close()
+
+		err = json.NewDecoder(resp.Body).Decode(&result)
+
+		fmt.Println(result)
+
+		assert.NoError(t, err)
+		assert.Equal(t, expected[i], result)
```

Comment:
When you will remove sorting, you will have problem here with the order of Devices - in that case you can either ignore individual devices or provide mock DAO with constant devices order (map in golang does not provide constant ordering)